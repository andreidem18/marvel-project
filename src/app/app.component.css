.page-container {
    display: flex;
}

.heroes {
    width: 100%;
}

.favorites {
    display: none;
}

@media (min-width: 1114px) {
    .favorites {
        display: block;
        width: 25%;
    }
    .heroes {
        width: 75%;
    }
}
