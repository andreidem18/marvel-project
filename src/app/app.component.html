<app-navbar />

<div class="page-container">
  <main class="heroes">
    <heroes-list />
  </main>

  <aside class="favorites">
    <comics-favorites-list />
  </aside>

</div>

<shared-toast />
<shared-confirmation />

