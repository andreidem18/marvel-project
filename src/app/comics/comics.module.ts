import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { SharedModule } from '../shared/shared.module';

import { ComicDetailComponent } from './components/comic-detail/comic-detail.component';
import { FavoritesListComponent } from './components/favorites-list/favorites-list.component';
import { FavoriteItemComponent } from './components/favorite-item/favorite-item.component';



@NgModule({
  declarations: [
    ComicDetailComponent,
    FavoritesListComponent,
    FavoriteItemComponent,
  ],
  imports: [
    CommonModule,

    SharedModule,
  ],
  exports: [
    ComicDetailComponent,
    FavoritesListComponent,
  ]
})
export class ComicsModule { }
