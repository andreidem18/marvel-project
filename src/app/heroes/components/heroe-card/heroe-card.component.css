.card-container {
    background: var(--surface-light);
    height: 100%;
}

.card-container .card-upper {
    display: flex;
    /* margin-left: -15px; */
    position: relative;
    left: -15px;
    top: -13px;
}

.card-container .card-upper .image-container {
    width: 45%;
    padding-right: 10px;
}

.card-container .card-upper .image-container img {
    border-radius: 50%;
    position: relative;
    object-fit: cover;
    aspect-ratio: 1 / 1;
    box-shadow: 6px 6px 5px -4px rgba(0,0,0,0.18);
}

.card-container .card-upper .hero-info {
    width: 55%;
    padding-left: 10px;
    display: flex;
    flex-direction: column;
    align-items: flex-start;
}

.hero-info .hero-name {
    display: block;
    color: var(--font-color-bold);
    margin-bottom: 8px;
    font-size: 1.2rem;
}

.hero-info .hero-description {
    font-size: .9rem;
    line-height: 1.3rem;
    color: var(--font-color-light);
    margin-bottom: 20px;
}

.hero-info .button-container {
    flex: 1;
    display: flex;
    align-items: flex-end;
    padding-bottom: 15px;
}

.hero-info button {
    background: var(--primary);
    color: #fff;
    padding: 10px;
    font-weight: 600;
    text-transform: uppercase;
}

.related-comics {
    padding: 0 20px 20px 20px;
}

.related-comics * {
    color: var(--font-color-light);
}

.related-comics > strong {
    display: block;
    width: fit-content;
    margin-bottom: .7rem;
}

.related-comics .related-comics-list {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    gap: .7rem;
}

.related-comics-list li {
    font-size: .85rem;
    cursor: pointer;
    display: flex;
    align-items: center;
}

.related-comics-list li span {
    padding-right: 14px;
}

.related-comics-list li img {
    width: 0;
    transition: .5s;
}

.related-comics-list li:hover span {
    color: var(--primary);
    padding-right: 0;
    padding-left: 3px;
    transition: .5s;
}

.related-comics-list li:hover img {
    width: 16px;
}

.related-comics .show-comics {
    display: block;
    width: fit-content;
    font-size: .9rem;
    margin-top: .7rem;
    margin-bottom: 1rem;
}

.related-comics .show-comics:hover {
    color: var(--primary);
    text-decoration: underline;
    cursor: pointer;
}



