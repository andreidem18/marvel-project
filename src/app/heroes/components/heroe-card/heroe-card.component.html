<div class="card-container">
    <div class="card-upper">
        <div class="image-container">
            <img [src]="heroImage" [alt]="hero.name">
        </div>
        <div class="hero-info">
            <strong class="hero-name" data-test="hero-name">{{hero.name}}</strong>
            <p class="hero-description">
                {{(hero.description ? hero.description : defaultDescription) | truncateText:150}}
            </p>
            <div class="button-container">
                <button (click)="openHeroDetail()">View more</button>
            </div>
        </div>
    </div>

    <div class="related-comics">
        <strong>Related comics</strong>
        <ul class="related-comics-list">
            <li 
                *ngFor="let comic of hero.comics.items.slice(0, 4)" 
                (click)="selectComic(comic)"
                data-test="comic-item"
            >
                <img src="../../../../assets/icons/link-arrow.png" alt="">
                <span>{{comic.name | truncateText:70}}</span>
            </li>
        </ul>
        <span 
            *ngIf="hero.comics.items.length > 4"
            class="show-comics"
            (click)="toggleShowAllComics()"
        >
            {{showAllComics ? '▲ Hide all comics' : '▼ Show all comics'}}
        </span>
        <ul class="related-comics-list" *ngIf="showAllComics">
            <li 
                *ngFor="let comic of hero.comics.items.slice(4)" 
                (click)="selectComic(comic)"
                data-test="comic-item"
            >
                <img src="../../../../assets/icons/link-arrow.png" alt="">
                <span>{{comic.name | truncateText:70}}</span>
            </li>
        </ul>
    </div>

</div>
