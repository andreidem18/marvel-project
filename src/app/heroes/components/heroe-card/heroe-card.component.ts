import { Component, EventEmitter, Input, OnInit, Output, inject } from '@angular/core';
import { ComicsItem, Hero } from '../../interfaces/Hero';
import { ToastService } from '../../../shared/services/toast/toast.service';

@Component({
  selector: 'heroe-card',
  templateUrl: './heroe-card.component.html',
  styleUrls: ['./heroe-card.component.css']
})
export class HeroeCardComponent implements OnInit {

  @Input()
  public hero!: Hero;

  @Output()
  public onSelectComic = new EventEmitter<ComicsItem>();

  private toastService = inject(ToastService);

  public showAllComics: boolean = false;

  public defaultDescription: string = 'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Doloremque voluptate nobis quas nesciunt atque illum aliquam incidunt sit veniam, asperiores dolorem, vero culpa accusantium sed id minus officia modi necessitatibus!'

  get heroImage(): string {
    const { extension, path } = this.hero.thumbnail
    return `${path}.${extension}`;
  }

  ngOnInit(): void {
    if(!this.hero) throw "You must send a valid hero!";
  }

  toggleShowAllComics() {
    this.showAllComics = !this.showAllComics;
  }

  selectComic(comic: ComicsItem) {
    this.onSelectComic.emit(comic);
  }

  openHeroDetail() {
    this.toastService.addToast({
      message: "this feature haven't been added yet",
      duration: 3,
      severity: 'warning',
    })
  }

}
