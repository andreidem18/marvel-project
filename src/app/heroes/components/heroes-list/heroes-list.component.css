.heroes-container {
    padding: 0 20px;
    padding-bottom: 5rem;
}

.input-search-container {
    margin-top: 1.5rem;
}

.header  {
    margin-top: 1.3rem;
    margin-bottom: 30px;
}

.header .characters-title {
    display: flex;
    align-items: center;
    margin-bottom: .5rem;
}

.header .characters-title img {
    width: auto;
}

.header .spinner {
    width: 80px;
    margin: 0 auto;
}

.header select {
    border: 1px solid #c0c0c0;
    border-radius: 1px;
    padding: 15px 15px;
    display: block;
    font-size: 1rem;
    text-indent: 5px;
    width: 100%;
}


.heroes-list {
    display: grid;
    grid-template-columns: 1fr;
    gap: 2rem;
    padding-bottom: 3rem;
}

.heroe-item {
    width: 100%;
    padding-left: 15px;
    padding-top: 13px;
}

@media (min-width: 570px) {
    .header {
        display: flex;
        justify-content: space-between;
    }
    .header select {
        width: 240px;
    }
}

@media (min-width: 768px) {
    .input-search-container {
        display: none;
    }
    .heroes-list {
        grid-template-columns: repeat(2, 1fr);
    }
}

