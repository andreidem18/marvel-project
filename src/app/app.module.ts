import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { HttpClientModule } from '@angular/common/http';

import { AppComponent } from './app.component';

import { ComicsModule } from './comics/comics.module';
import { HeroesModule } from './heroes/heroes.module';
import { SharedModule } from './shared/shared.module';
import { httpInterceptorProviders } from './http-interceptors';
import { ReactiveFormsModule } from '@angular/forms';
import { NavbarComponent } from './standalone/navbar/navbar.component';

@NgModule({
  declarations: [
    AppComponent,
  ],
  imports: [
    BrowserModule,
    HttpClientModule,
    ReactiveFormsModule,

    ComicsModule,
    HeroesModule,
    NavbarComponent,
    SharedModule,
  ],
  providers: [
    httpInterceptorProviders,
  ],
  bootstrap: [AppComponent]
})
export class AppModule { }
