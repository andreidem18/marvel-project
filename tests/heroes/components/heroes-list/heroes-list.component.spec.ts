import { ComponentFixture, TestBed } from '@angular/core/testing';
import { heroesResMock } from '../../mocks-data/heroes-res.mock';

import { HeroesListComponent } from '../../../../src/app/heroes/components/heroes-list/heroes-list.component';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { HeroesService } from 'src/app/heroes/services/heroes.service';
import { environtment } from 'src/environments/environments';
import { httpInterceptorProviders } from 'src/app/http-interceptors';
import { ComicsModule } from 'src/app/comics/comics.module';
import { SharedModule } from 'src/app/shared/shared.module';
import { ReactiveFormsModule } from '@angular/forms';
import { HeroeCardComponent } from 'src/app/heroes/components/heroe-card/heroe-card.component';

describe('HeroesListComponent', () => {
  let apiUrl = environtment.apiUrl;

  let component: HeroesListComponent;
  let fixture: ComponentFixture<HeroesListComponent>;
  let compiled: HTMLElement;
  let service: HeroesService;
  let httpMock: HttpTestingController;


  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [HeroesListComponent, HeroeCardComponent],
      imports: [HttpClientTestingModule, ComicsModule, SharedModule, ReactiveFormsModule],
    });
    fixture = TestBed.createComponent(HeroesListComponent);
    component = fixture.componentInstance;
    service = TestBed.inject(HeroesService);
    httpMock = TestBed.inject(HttpTestingController);
    
    const request = httpMock.expectOne(`${apiUrl}/characters?offset=0&limit=10`);
    request.flush(heroesResMock);

    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match with the snapshot', () => {
    expect(compiled.innerHTML).toMatchSnapshot();
  });


  it('should load a list of heroes', () => {

    const heroesCards = compiled.querySelectorAll('.heroe-item');

    expect(heroesCards).toHaveLength(heroesResMock.data.results.length);
  });

  it('should call a request when changing order by', () => {
    const select = compiled.querySelector('select');

    
    expect(select).not.toBeNull();

    select!.value = 'modified';
    select!.dispatchEvent(new Event('change'));
    httpMock.match(`${apiUrl}/characters?offset=0&limit=10&orderBy=modified`);
  });

  it('should call a request when executing setPagination', () => {
    component.setPagination({limit: 10, offset: 30});
    httpMock.expectOne(`${apiUrl}/characters?offset=30&limit=10`)
  })

  
  
});
