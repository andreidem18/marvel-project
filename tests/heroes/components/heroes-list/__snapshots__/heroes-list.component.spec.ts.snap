// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroesListComponent should match with the snapshot 1`] = `
"<section class="heroes-container"><div class="input-search-container"><shared-input-search><input type="text" placeholder="Search a character" class="input-search ng-untouched ng-pristine ng-valid" ng-reflect-form="[object Object]"></shared-input-search></div><div class="header"><div class="characters-title"><img src="../../../../assets/icons/characters.png" alt=""><strong>Characters</strong></div><!--bindings={
  "ng-reflect-ng-if": "false"
}--><select ng-reflect-form="[object Object]" class="ng-untouched ng-pristine ng-valid"><option value="name" ng-reflect-value="name">Sort by</option><option value="name" ng-reflect-value="name">Name</option><option value="modified" ng-reflect-value="modified">Modified date</option></select></div><ul class="heroes-list"><li class="heroe-item"><heroe-card ng-reflect-hero="[object Object]"><div class="card-container"><div class="card-upper"><div class="image-container"><img src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg" alt="3-D Man"></div><div class="hero-info"><strong data-test="hero-name" class="hero-name">3-D Man</strong><p class="hero-description"> Lorem ipsum dolor sit amet, consectetur adipisicing elit. Doloremque voluptate nobis quas nesciunt atque illum aliquam incidunt sit veniam, asperiores... </p><div class="button-container"><button>View more</button></div></div></div><div class="related-comics"><strong>Related comics</strong><ul class="related-comics-list"><li data-test="comic-item"><img src="../../../../assets/icons/link-arrow.png" alt=""><span>Avengers: The Initiative (2007) #14</span></li><li data-test="comic-item"><img src="../../../../assets/icons/link-arrow.png" alt=""><span>Avengers: The Initiative (2007) #14 (SPOTLIGHT VARIANT)</span></li><li data-test="comic-item"><img src="../../../../assets/icons/link-arrow.png" alt=""><span>Avengers: The Initiative (2007) #15</span></li><li data-test="comic-item"><img src="../../../../assets/icons/link-arrow.png" alt=""><span>Avengers: The Initiative (2007) #16</span></li><!--bindings={
  "ng-reflect-ng-for-of": "[object Object],[object Object"
}--></ul><span class="show-comics"> ▼ Show all comics </span><!--bindings={
  "ng-reflect-ng-if": "true"
}--><!--bindings={
  "ng-reflect-ng-if": "false"
}--></div></div></heroe-card></li><li class="heroe-item"><heroe-card ng-reflect-hero="[object Object]"><div class="card-container"><div class="card-upper"><div class="image-container"><img src="http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16.jpg" alt="A-Bomb (HAS)"></div><div class="hero-info"><strong data-test="hero-name" class="hero-name">A-Bomb (HAS)</strong><p class="hero-description"> Rick Jones has been Hulk's best bud since day one, but now he's more than a friend...he's a teammate! Transformed by a Gamma energy explosion, A-Bomb'... </p><div class="button-container"><button>View more</button></div></div></div><div class="related-comics"><strong>Related comics</strong><ul class="related-comics-list"><li data-test="comic-item"><img src="../../../../assets/icons/link-arrow.png" alt=""><span>FREE COMIC BOOK DAY 2013 1 (2013) #1</span></li><li data-test="comic-item"><img src="../../../../assets/icons/link-arrow.png" alt=""><span>Hulk (2008) #53</span></li><li data-test="comic-item"><img src="../../../../assets/icons/link-arrow.png" alt=""><span>Hulk (2008) #54</span></li><li data-test="comic-item"><img src="../../../../assets/icons/link-arrow.png" alt=""><span>Hulk (2008) #55</span></li><!--bindings={
  "ng-reflect-ng-for-of": "[object Object],[object Object"
}--></ul><!--bindings={
  "ng-reflect-ng-if": "false"
}--><!--bindings={
  "ng-reflect-ng-if": "false"
}--></div></div></heroe-card></li><!--bindings={
  "ng-reflect-ng-for-of": "[object Object],[object Object"
}--></ul><shared-pagination ng-reflect-limit="20" ng-reflect-offset="0" ng-reflect-total="1562"><ul class="pagination-container"><li class="pagination-button center button disabled" ng-reflect-ng-class="[object Object]"> ⮜ </li><li class="pagination-button center selected" ng-reflect-ng-class="[object Object]"> 1 </li><li class="pagination-button center" ng-reflect-ng-class="[object Object]"> 2 </li><li class="pagination-button center" ng-reflect-ng-class="[object Object]"> 3 </li><li class="pagination-button center" ng-reflect-ng-class="[object Object]"> 4 </li><li class="pagination-button center" ng-reflect-ng-class="[object Object]"> 5 </li><!--bindings={
  "ng-reflect-ng-for-of": "1,2,3,4,5"
}--><li class="pagination-button center button" ng-reflect-ng-class="[object Object]"> ⮞ </li></ul></shared-pagination></section><!--bindings={
  "ng-reflect-ng-if": "[object Object]"
}--><comic-detail ng-reflect-is-open="false"><shared-modal-wrapper style="width: 80%;" ng-reflect-is-open="false"><div class="modal-container wrap center" ng-reflect-ng-class="[object Object]"><div class="modal-body"><!--bindings={}--></div><div class="overlay wrap"></div></div></shared-modal-wrapper></comic-detail>"
`;
